### ISSUE TITLE
Add `include_all` optarg to `.changes()`

### ISSUE THREAD

**Issue by marshall007** [rethinkdb/rethinkdb#3719]
Discussion started with [my comment](https://github.com/rethinkdb/rethinkdb/issues/3709#issuecomment-72607276) over in #3709. There are a lot of cases where the overhead of returning the entire result set on each change would be worth it for the sake of simplicity. For example if I want the top 5 users by score the performance impact of emitting all five rows on every change is minimal and I'd like to avoid writing the code necessary to keep my in-memory copy in sync/ordered correctly.

**Comment by danielmewes** [comment]
This generally sounds like a good idea to me.

An open question is what exactly the changefeed should return when the data changes.
Assuming an `orderBy(...).limit(...)` changefeed, it could have `old_val` and `new_val` be arrays with all the documents in the right order.
In that case we should probably send the initial values in the same format, that is as a single document `{new_val: [val1, val2, val3, ...]}`.

I think for now the only query that this would work on is `orderBy.limit`?
Even if we support `return_initial` for other multi-document changefeeds (like `r.table().changes()` or `r.table().between().changes()`) later, I don't think we could easily (let alone efficiently) support this opt arg for those types of changefeeds, could we?
This makes this solution seem a bit specific, and I wonder wonder if there's a more generic solution to the problem. Though I cannot think of any.

**Comment by timmaxw** [comment]
I suggest the syntax `r.table(...).orderBy(...).limit(...).coerceTo("ARRAY").changes()`. Then it's obvious that the first message should be `{new_val: [val1, val2, val3, ...]}` and subsequent messages should be the same thing with `old_val` too.

**Comment by danielmewes** [comment]
Uh I like that @timmaxw .

**Comment by mlucy** [comment]
I really don't like that, personally.
* It's much less discoverable than an optarg to changes.
* It doesn't make much sense in the context of our type system.
  - Why does a stream send diffs of the set and an array send the whole set?
  - Why can I only call `changes` on arrays produced by this one query?  (We currently have the rule that you can call `changes` on all selections and some streams in 2.0.)
* It's more work to implement (partially because of the above point about it not really lining up with our type system -- there's currently no place to attach changefeed information to datum arrays).

---

I think we should just add an optarg to changes.  I think it should work on `.orderby.limit.changes`, `.between.changes`, and `.get_all.changes`, with the caveat that if you use it with `between` or `get_all` we limit the size of the result set to the array size limit (which we already do for `.orderby.limit`).

**Comment by mlucy** [comment]
We may also want to consider making the result with the optarg take the form `{new_vals: ..., old_vals: ...}` so that it's obvious you're getting back something different.

**Comment by marshall007** [comment]
@mlucy I can imagine scenarios where you'd want to generically emit changes to the client and pluralizing `new_val/old_val` in certain cases would complicate things. If anything I would do `{new_val: ..., old_val: ..., includes_all: true}`

**Comment by danielmewes** [comment]
We could also make the `includes_all: true` annotation a flag in the `ResponseNotes` field proposed here https://github.com/rethinkdb/rethinkdb/issues/3715#issuecomment-72765917.

**Comment by timmaxw** [comment]
I agree with @mlucy that it should be an optarg. But I think it should work for both changefeeds on streams and changefeeds on arrays. For example, if you do `r.table(...).map(...).reduce(...)` and the result is an array, you should be able to call `.changes(includes_all = False)` to have the server diff your arrays and send you changes in the same format as `r.table(...).changes()`.

**Comment by mlucy** [comment]
I talked about this with @timmaxw offline and I understand where he was coming from with the `.coerce_to('ARRAY')` syntax.  I agree that in the future when we have changefeeds on more things being able to send diffs on datums will be a good idea.

**Comment by mlucy** [comment]
(I.e. I still think we should use an optarg for this particular task, but changefeeds on arrays no longer seem crazy to me.)

**Comment by danielmewes** [comment]
Since it seems easier to implement, I agree with @mlucy that we should do the optarg for this one.

When we support changefeeds on a query coerced to an array at some point, we can consider deprecating it or make it a rewrite rule.

**Comment by mlucy** [comment]
Planning to mark this as settled on Monday unless anyone objects, with the following interpretation:
* We will add an optarg `include_all` to `changes` which will default to `false` for all currently-existing changefeeds.
* Setting this optarg to `true` will error except on `.orderby.limit.changes` (in the future we may also support it for small ranges, or for `get_all`).  Setting this optarg to `true` will cause the server to re-send the entire ordered set every time something changes.
* The format will be identical except that `new_val` and `old_val` will be arrays of documents rather than single documents.  We will add a `ResponseNote` (see https://github.com/rethinkdb/rethinkdb/issues/3715) indicating that we're using a variant format.

**Comment by marshall007** [comment]
After thinking about this a bit more, it might be more consistent if the result were an array containing new/old value pairs i.e. `[ {new_val: ..., old_val: ...}, {new_val: ..., old_val: ...} ]`. That way you don't have to manually zip the two arrays together before passing them to your code that handles the existing changefeed behavior. I don't have strong opinions either way, though.

**Comment by danielmewes** [comment]
@marshall007 Can you explain a bit more what you mean by having to zip the arrays together? I'm not sure I understand what you mean.
I imagined the idea was that you could just overwrite the whole data with a set of new values, which you would find in `new_val` when `include_all` is set.

I think having the individual new/old value pairs is a little awkward because most of them will usually be the same with `include_all`.
Having two arrays `{old_val: [...], new_val: [...]}` looks a bit nicer to me.

**Comment by mlucy** [comment]
I think I would probably prefer to get `{old_val: [...], new_val: [...]}` as a user; like @danielmewes said, I think the primary use case for this is when you just want to blow away your old set and substitute in the new one.

**Comment by marshall007** [comment]
> I imagined the idea was that you could just overwrite the whole data with a set of new values, which you would find in `new_val` when `include_all` is set.

For the use case I originally described that is correct. However, if your code cares about comparing the new/old values (for sanity checks, audit logging, etc) its nice to be able to write a generic function that takes `{ new_val, old_val }` and can be used in both cases.

If all I want to do is blow away my old set I can just do `r.table(...).changes({ include_all: true })('new_val')` so what I proposed still supports both use cases.

Also for the `.getAll(...)` case you'd end up with sparsely populated arrays when rows are deleted/added which seems annoying.

**Comment by danielmewes** [comment]
@marshall007 Hmm I can see that the `{old_val: [], new_val: []}` format makes it more difficult to diff in cases where values are added or removed to/from the result set.
I still like the this format better because it avoids pairs of `old_val` and `new_val` that have the same value.

When you talk about the sparse array, do you mean the `[{old_val: ..., new_val: ...}, {old_val: ..., new_val: ...}, ...]` syntax? I actually don't think it's a huge problem if some of the vale will be `null`.
In the `{old_val: [...], new_val: [...]}` case the arrays would never be sparse in any sense I can think of, would it?

**Comment by marshall007** [comment]
> In the `{old_val: [...], new_val: [...]}` case the arrays would never be sparse in any sense I can think of, would it?

Yea that's the syntax I was talking about. Imagine that the table is initially empty and I'm watching on `r.table(...).getAll(...).changes({ include_all: true })`. Once a document is inserted that matches the `getAll` wouldn't I get `{old_val: [ null ], new_val: [ { ... } ]}`? So `old_val` is sparse.

**Comment by timmaxw** [comment]
I don't think we should make the arrays sparse in that case. It's too confusing for users who just want to have a list of values that gets updated, and don't want to do any processing beyond extracting `new_val` and sending it out to the client.

**Comment by marshall007** [comment]
> I still like the this format better because it avoids pairs of `old_val` and `new_val` that have the same value.

And to address this part, like I said in my previous comment, if you don't care about diffing the values you should be using `.changes({ include_all: true })('new_val')` to avoid getting the extra data all together.

**Comment by danielmewes** [comment]
Ok there seem to be three possible variants here:

1. return `[{old_val: ..., new_val: ...}, {old_val: ..., new_val: ...}, ...]`. Unchanged values will have both old_val and new_val set to the same value.
  
  Pro:
  - you still get a diff
  
  Con:
  - we get lots of pairs where `old_val` is the same as `new_val`. We don't generate such pairs anywhere else, so this would be a bit of an inconsistency.
  - to overwrite the result set in your application completely, you need to append a `.filter(r.row.ne(null))` after `changes`
2. return `{old_val: [v1, v2, v3, ...], new_val: [v1', v2', v3', ...]}`. If a value gets removed it will not appear in `new_val`. If it gets added it will not appear in `old_val` (i.e. there won't be any `null`s in the arrays)
  
  Pro:
  - trivial to use if you just want to update the full result
  
  Con:
  - no easy way to still get a diff (e.g. if you want to highlight changes in your UI)
3. return `{old_val: [v1, v2, v3, ...], new_val: [v1', v2', v3', ...]}`. If a value gets removed put a `null` in its place in `new_val`, if it gets added put one in `old_val`.
  
  Pro:
  - you still get a diff
  
  Con:
  - the diff comes in a different format than usual
  - to overwrite the result set in your application completely, you need to append a `.filter(r.row.ne(null))` after `changes`


I really like option 2 the most, since it's very simple. I feel like 1 is less intuitive and in some cases more difficult to use. Though being able to get the diff is certainly an advantage.

**Comment by timmaxw** [comment]
I also prefer option 2. The other options make the most common case more complicated.

**Comment by mlucy** [comment]
Option 2 seems good to me as well.

**Comment by mlucy** [comment]
Marking as settled for option 2 since we seem to have a quorum.

**Comment by marshall007** [comment]
Sounds good, and to clarify will `.get_all.changes` be supported for small ranges in the initial implementation or just `.orderby.limit.changes`?

**Comment by mlucy** [comment]
I'm not sure, actually.  `.get_all.changes` is in https://github.com/rethinkdb/rethinkdb/issues/3642 and doesn't have any discussion of whether it will support `return_initial`.  It probably depends on how much time we have before the release.

**Comment by coffeemug** [comment]
I'm sorry to bring this up now, as this issue seems to be settled, but since we're going through another reql proposal period, I'd like share an opinion on this API. I think everything decided here is great, but option 2 might be a bad choice.

Firstly, as @danielmewes pointed out you can't get a diff. Since you can't get a diff, returning `old_vals` seems kind of pointless -- there isn't really anything the user can do with that data. Additionally, the API is a little confusing because it appears the actual values are arrays. We're sort of shoehorning the new functionality into an old API in a way that IMO (a) isn't very useful, and (b) confusing/inelegant.

I'd like to propose a different option.

Return `[{old_val: ..., new_val: ...}, {old_val: ..., new_val: ...}, ...]`. Unchanged values will have only `old_val`, but no `new_val` (as is, the `new_val` field would be missing). We'd also have to include the state (or possibly force people to set `include_state` to true if they set `include_all` to true), because otherwise the user couldn't tell when the full change begins and ends.

I think this API is *slightly* harder to use, but is more general, more useful, and is still pretty easy to explain.

We could also consider other options, I just think option 2 is a little bit sloppy.

**Comment by coffeemug** [comment]
A few more things:

- For people reading over email, I edited the proposal above, please see on github.
- @danielmewes pointed out that you couldn't just say `...changes(include_all=True)['new_val]` to replace everything. This is pretty bad, so perhaps we should always include both `old_val` and `new_val`, and eat the cost of additional traffic.

**Comment by mlucy** [comment]
We usually have a policy of not-unsettling things, but since @coffeemug didn't get to comment during the original discussion for whatever reason I think it's OK here (as long as we don't make a habit of it).

What do people think of the available options after looking at them with fresh eyes?

**Comment by mlucy** [comment]
After reflecting on this:
* I still like option 2 the best.
* I now think the name `include_all` is terrible.  We've been using the `include_` prefix (for `include_state` and the `include_initial_vals` in CR) to mean "add more information to the stream", not "change the format of the information in the stream".  I can imagine users being confused about why `include_all` doesn't cause e.g. state documents to be included.
  - I think we should instead have an optarg `format` which can be set to either `diff` or `set` (or possibly better names), with `diff` being the default and `set` being the format described in option 2.

**Comment by coffeemug** [comment]
Apologies again for asking to reopen.

I think that if the optarg can take on the values of `diff`, and `set`, option 2 makes a lot more sense (heh, looks like it *is* all in the naming :smile:). I'd consider finding a better name for `set`, though. Maybe `full`? Also, `format` is a little confusing because we normally use that to control the representation of pseudotypes (also, see #2967). May be we should call the optarg `feed_type`?

However, the format of option two still seems strange to me. What would be the point of returning the old values? Since the user can't correlate them with new values, they essentially can't do anything with that data -- we might as well just send an array of elements for each change, and drop the `old_val`/`new_val` fields altogether.

**Comment by mlucy** [comment]
* `full` instead of `set` sounds good to me.
* I like `format` better than `feed_type`, but I don't really like either.  Maybe we can come up with a better one?  What do other people think?
* I wouldn't be opposed to the variant of option 2 where we just return a stream of arrays; I think you're right that `old_val` is vestigal with that format.

**Comment by coffeemug** [comment]
How about `output=diff | full`?

**Comment by danielmewes** [comment]
Option 2 as a sequence of arrays without an `old_val` sounds good.

I don't like `feed_type`, and `format` (which I do like otherwise) has the mentioned conflict with the pseudo type format arguments. `output` would be ok, but I don't think it's perfect.

What do you (and others) think about something more similar to the original `include_all` just without `include`? For example `resend_all`, `resend_full` or `send_full`?

**Comment by coffeemug** [comment]
I'd prefer an optarg with multiple values, rather than a boolean optarg. It's subjective, but it feels dramatically nicer to me.

We can think for a bit and try to come up with a better name than `output`, but if we can't think of a better name, I think `output` is good enough.

**Comment by marshall007** [comment]
I don't want to veer too far off course here but another option would be kind of a mix of what @coffeemug suggested and option (1) from @danielmewes comment. To have an optarg called `exclude` or `omit` with the following options:

- `inserted` - drop rows with no `old_val`
- `deleted` - drop rows with no `new_val`
- `unchanged` - drop rows with either no `old_val` or `new_val`
- `replaced` - drop rows that changed (not sure why you'd want this, I just added it for completion)
- `false (default)` - traditional changefeed behavior i.e. option (1)

I think this supports all mentioned use-cases and then some including the `changes(include_all=True, omit='deleted')['new_val]` for either side.

**Comment by mlucy** [comment]
I agree with @coffeemug; I like having two non-boolean values.  `diff` and `full` seem fine to me.  How about `feed_format` for the name of the arg?  (If that doesn't work, I think `output` is my favorite so far.)

**Comment by danielmewes** [comment]
`feed_format` is similar to `format`. I quite like it, but it also follows the same scheme as `binary_format`, `group_format` and `time_format`.

@marshall007 That's actually an interesting way of looking at this. So the default would be omitting `unchanged`, and what we're discussing here would be equivalent to including those. I think it's a bit too complicated though, and the order in which `unchanged` "changes" arrive relative to the actual changes isn't quite obvious. Currently it looks like we'll be sending a complete array with all values as a single change, which I think is easier to understand and use.

**Comment by marshall007** [comment]
Ok so to clarify/summarize if we go with the `feed_format=diff|full`

- `diff`: implements @coffeemug's proposal. i.e. `[{old_val: ..., new_val: ...}, {old_val: ..., new_val: ...}, ...]` and unchanged values will only have `old_val` populated
- `full`: implements option (2). i.e. `{old_val: [v1, v2, v3, ...], new_val: [v1', v2', v3', ...]}`. If a value is removed it will not appear in `new_val`, If it's added it will not appear in `old_val`

My only gripe with this is that in the proposed `diff` format it's still impossible to disambiguate between deleted and unchanged values. That seems like an important distinction for some use-cases. I don't see a good reason (aside from overhead) not to populate both `old_val` and `new_val` in the unchanged case.

**Comment by mlucy** [comment]
@marshall007 -- I think the idea is that `diff` is the current behavior (unchanged values aren't sent at all), and `full` is the behavior you described.

**Comment by coffeemug** [comment]
That wasn't my impression. I thought:

- `diff`: works like it does now. Unchanged values don't show up at all (it's a diff!)
- `full`: sends a stream of arrays, e.g. `[v1, v2, v3]` (we talked above about how including `old_val` at all would be completely superfluous with this option).

**Comment by marshall007** [comment]
Got it, I like that. It's a lot cleaner than what I thought was being suggested.

**Comment by coffeemug** [comment]
@marshall007 -- could you shoot me an e-mail at slava@rethinkdb.com? I wanted to ask some questions about the use cases offline -- it would help immensely with feature planning.

**Comment by marshall007** [comment]
@coffeemug you bet, just emailed you.

**Comment by danielmewes** [comment]
Marking as settled.
We'll add an `output` optarg that can take the values `diff` or `full`. The default is `diff`, which is equivalent to the current behavior. Setting it to `full` results in the changefeed producing a stream of arrays `[v1, v2, v3]`.

**Comment by coffeemug** [comment]
:+1:

**Comment by coffeemug** [comment]
I've been speaking with many people about this recently, and learned that people really want the notion of json diffs for feeds (because if a tiny part of the document changes, they don't want us to send the whole document, but just the diff).

With that in mind, I'd like to augment the proposal -- instead of `output=diff | full`, I propose we do `output=point | full` (basically, just rename `diff` to `point`). We can then add `diff` later that will do the *actual* diff.

Also, I think this should be required for 2.2, as without it `orderBy` feeds are basically unusable (my impression was that we agreed to include this with the other feed changes to significantly bump up usability).

**Comment by danielmewes** [comment]
> instead of `output=diff | full`, I propose we do `output=point | full`

How about `output = document | full`? I think `point` is rather unspecific.

**Comment by coffeemug** [comment]
That sounds good.

**Comment by coffeemug** [comment]
Or maybe `single`, `all`, and eventually `diff`.

**Comment by danielmewes** [comment]
I like `single`, `all` + `diff`

**Comment by coffeemug** [comment]
Note, the issue for `diff` format is #4707.
